src/theory/uf/distinct_extension.cpp

```diff
@@ -0,0 +1,252 @@
+                ? new EagerProofGenerator(d_env, context(), "DistinctEpg")
+                : nullptr),
+      d_pendingConflict(context())
+{
+  d_false = nodeManager()->mkConst(false);
+}
+
+bool DistinctExtension::needsCheckLastEffort()
+{
+  return d_negDistinctIndex.get() < d_negDistinct.size()
+         || !d_posDistinct.empty();
+}
+
+void DistinctExtension::assertDistinct(TNode atom, bool pol, TNode fact)
+{
+  for (const Node& nc : atom)
+  {
+    d_state.addTerm(nc);
+  }
+  if (pol)
+  {
+    d_posDistinct.push_back(fact);
+    std::unordered_map<Node, Node> reps;
+    std::unordered_map<Node, Node>::iterator itr;
+    bool isConflict = false;
+    for (const Node& nc : atom)
+    {
+      Node ncr = d_state.getRepresentative(nc);
+      itr = reps.find(ncr);
+      if (itr == reps.end())
+      {
+        reps[ncr] = nc;
+        continue;
+      }
+      isConflict = true;
+      // otherwise already a conflict
+      Node eq = itr->second.eqNode(nc);
+      d_im.conflictExp(InferenceId::UF_DISTINCT_DEQ, {eq, fact}, d_epg.get());
+      break;
+    }
+    if (!isConflict)
+    {
+      for (const std::pair<const Node, Node>& p : reps)
+      {
+        Trace("uf-lazy-distinct")
+            << "Watch " << p.first << " distinct (" << fact << ")" << std::endl;
+        size_t ndprev = d_ndistinct[p.first];
+        d_ndistinct[p.first] = ndprev + 1;
+        // ensure the non-context dependent list has the right size in
+        // case we backtracked
+        std::vector<Node>& ndlist = d_eqcToDistinct[p.first];
+        ndlist.resize(ndprev);
+        ndlist.emplace_back(fact);
+        // also carry the member
+        std::vector<Node>& ndmem = d_eqcToDMem[p.first];
+        ndmem.resize(ndprev);
+        ndmem.emplace_back(p.second);
+      }
+    }
+  }
+  else
+  {
+    // otherwise track that the negated distinct term
+    d_negDistinct.push_back(fact);
+  }
+}
+
+void DistinctExtension::eqNotifyMerge(TNode t1, TNode t2)
+{
+  // Must ensure we track distinct constraints, moving those from t2 into t1.
+  // If the same distinct constraint is in both, we are in conflict.
+  NodeUIntMap::iterator it2 = d_ndistinct.find(t2);
+  if (it2 != d_ndistinct.end())
+  {
+    Trace("uf-lazy-distinct") << "merge " << t1 << " and " << t2 << std::endl;
+    NodeUIntMap::iterator it1 = d_ndistinct.find(t1);
+    std::vector<Node>& d1 = d_eqcToDistinct[t1];
+    std::vector<Node>& d1m = d_eqcToDMem[t1];
+    std::vector<Node>& d2 = d_eqcToDistinct[t2];
+    // the iterator up to which d2 is valid
+    std::vector<Node>::iterator d2e = d2.begin() + it2->second;
+    if (it1 != d_ndistinct.end())
+    {
+      // ensure the list of distinct constraints in t1 is resized now
+      d1.resize(it1->second);
+      d1m.resize(it1->second);
+      Trace("uf-lazy-distinct")
+          << "...looking for conflicts in intersection of # terms = "
+          << it1->second << " and " << it2->second << std::endl;
+      // check for conflicts
```

Comment:
clarify how this check is achieving what is described in the beginning of the method